File: 백준/Gold/17298. 오큰수/오큰수.cc
#include<iostream>
#include<stack>
#include<cstring>
#include<vector>
#define FAST_IO cin.tie(0), cout.tie(0), ios_base::sync_with_stdio(0)
#define F first
#define S second
using namespace std;
typedef pair<int, int> PII;

int n;
vector<PII> dat;
stack<int> sta;



int main(void) {
	FAST_IO;
	cin >> n;

	
	// 4 -> 3 5 2 7
	for (int i = 0; i < n; i++) {
		int cur;
		cin >> cur;
		dat.push_back({ cur, -1 }); // 값, 현재 index로 저장
			
		while (!sta.empty() && dat[sta.top()].F < cur) {
			dat[sta.top()].S = cur;
			sta.pop();
		}
		sta.push(i);
	}

	for (int i = 0; i < n; i++) cout << dat[i].S << " ";
	return 0;
}

File: 백준/Gold/17298. 오큰수/오큰수.py
import sys
input = sys.stdin.readline

n = int(input())
arr = list(map(int, input().split()))
dat = [[val, -1] for val in arr]
sta = []

for i in range(n):
    cur = arr[i]
    # 오큰수 처리
    while sta and dat[sta[-1]][0] < cur:
        dat[sta[-1]][1] = cur
        sta.pop()
    sta.append(i)

print(' '.join(str(dat[i][1]) for i in range(n)))
